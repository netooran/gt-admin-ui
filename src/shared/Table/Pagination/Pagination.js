import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  ChevronLeft,
  ChevronsLeft,
  ChevronRight,
  ChevronsRight,
} from '@styled-icons/boxicons-solid';
import Button from 'shared/Button/Button';
import ButtonGroup from 'shared/ButtonGroup/ButtonGroup';
import { ArrayUtils } from 'utils/utils/utils';

const StyledButton = styled(Button)`
  --button-primary-color: #1890ff;
`;

const Pagination = ({ page, noOfpages, goto, noOfPagesToDisplay = 5 }) => {
  const hasPrevious = page === 1;
  const hasNext = page === noOfpages;

  const getPages = () => {
    const offset = page > noOfPagesToDisplay ? page - noOfPagesToDisplay : 0;
    return ArrayUtils.range(noOfpages).slice(
      offset,
      offset + noOfPagesToDisplay
    );
  };

  return (
    <ButtonGroup>
      <StyledButton
        aria-label="Go to first page"
        onClick={() => goto(1)}
        disabled={hasPrevious}
        variant="solid"
        shape="round"
      >
        <ChevronsLeft />
      </StyledButton>
      <StyledButton
        aria-label="Go to previous page"
        onClick={() => goto(page - 1)}
        disabled={hasPrevious}
        variant="solid"
        shape="round"
      >
        <ChevronLeft />
      </StyledButton>
      {getPages().map((pageNumber, i) => (
        <StyledButton
          aria-label={`Go to page ${pageNumber}`}
          key={i}
          onClick={() => goto(pageNumber)}
          variant={page === pageNumber ? 'outline' : 'solid'}
          shape="round"
        >
          {pageNumber}
        </StyledButton>
      ))}
      <StyledButton
        aria-label="Go to next page"
        onClick={() => goto(page + 1)}
        disabled={hasNext}
        variant="solid"
        shape="round"
      >
        <ChevronRight />
      </StyledButton>
      <StyledButton
        aria-label="Go to last page"
        onClick={() => goto(noOfpages)}
        disabled={hasNext}
        variant="solid"
        shape="round"
      >
        <ChevronsRight />
      </StyledButton>
    </ButtonGroup>
  );
};

Pagination.propTypes = {
  page: PropTypes.number.isRequired,
  noOfpages: PropTypes.number.isRequired,
  noOfPagesToDisplay: PropTypes.number,
  goto: PropTypes.func.isRequired,
};

export default Pagination;
