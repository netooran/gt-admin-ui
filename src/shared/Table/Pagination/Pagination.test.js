import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Pagination from './Pagination';

const mockGoto = jest.fn();

const assertGotoPageToHaveBeencalled = (label, page, disabled = false) => {
  mockGoto.mockClear();

  userEvent.click(screen.getByLabelText(label));

  if (disabled) expect(mockGoto).not.toHaveBeenCalled();
  else expect(mockGoto).toHaveBeenCalledWith(page);
};

const assertGotoPageToNotHaveBeencalled = (...args) =>
  assertGotoPageToHaveBeencalled(...args, true);

test('renders pagination with 1st page active', () => {
  const props = {
    page: 1,
    noOfpages: 5,
    goto: mockGoto,
  };

  const { container } = render(<Pagination {...props} />);
  expect(container).toMatchSnapshot();

  assertGotoPageToNotHaveBeencalled('Go to first page', 1);
  assertGotoPageToNotHaveBeencalled('Go to previous page', 1);
  assertGotoPageToHaveBeencalled('Go to page 3', 3);
  assertGotoPageToHaveBeencalled('Go to next page', 2);
  assertGotoPageToHaveBeencalled('Go to last page', 5);
});

test('renders pagination with last page active', () => {
  const props = { page: 5, noOfpages: 5, goto: mockGoto };

  const { container } = render(<Pagination {...props} />);
  expect(container).toMatchSnapshot();

  assertGotoPageToHaveBeencalled('Go to first page', 1);
  assertGotoPageToHaveBeencalled('Go to previous page', 4);
  assertGotoPageToHaveBeencalled('Go to page 3', 3);
  assertGotoPageToNotHaveBeencalled('Go to next page', 5);
  assertGotoPageToNotHaveBeencalled('Go to last page', 5);
});

test('renders pagination with a page in the middle active', () => {
  const props = { page: 3, noOfpages: 5, goto: mockGoto };

  const { container } = render(<Pagination {...props} />);
  expect(container).toMatchSnapshot();

  assertGotoPageToHaveBeencalled('Go to first page', 1);
  assertGotoPageToHaveBeencalled('Go to previous page', 2);
  assertGotoPageToHaveBeencalled('Go to next page', 4);
  assertGotoPageToHaveBeencalled('Go to last page', 5);
});

test('renders pagination with a page greater than default pages to display (5)', () => {
  const props = { page: 6, noOfpages: 10, goto: mockGoto };
  const { container } = render(<Pagination {...props} />);
  expect(container).toMatchSnapshot();
});
