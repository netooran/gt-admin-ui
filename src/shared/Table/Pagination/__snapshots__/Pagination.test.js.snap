// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders pagination with 1st page active 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: var(--button-primary-color);
  color: white;
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c1 svg {
  width: 1.2rem;
}

.c1:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c1:hover {
  opacity: 1;
}

.c4 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c4 svg {
  width: 1.2rem;
}

.c4:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c4:hover {
  opacity: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 button {
  margin-right: 0.5rem;
}

.c2 {
  --button-primary-color: #1890ff;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-label="Go to first page"
      class="c1 c2"
      disabled=""
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.121 12l4.94-4.939-2.122-2.122L3.879 12l7.06 7.061 2.122-2.122z"
        />
        <path
          d="M17.939 4.939L10.879 12l7.06 7.061 2.122-2.122L15.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to previous page"
      class="c1 c2"
      disabled=""
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.939 4.939L6.879 12l7.06 7.061 2.122-2.122L11.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to page 1"
      class="c4 c2"
      shape="round"
    >
      1
    </button>
    <button
      aria-label="Go to page 2"
      class="c1 c2"
      shape="round"
    >
      2
    </button>
    <button
      aria-label="Go to page 3"
      class="c1 c2"
      shape="round"
    >
      3
    </button>
    <button
      aria-label="Go to page 4"
      class="c1 c2"
      shape="round"
    >
      4
    </button>
    <button
      aria-label="Go to page 5"
      class="c1 c2"
      shape="round"
    >
      5
    </button>
    <button
      aria-label="Go to next page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.061 19.061L17.121 12l-7.06-7.061-2.122 2.122L12.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to last page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.061 4.939l-2.122 2.122L15.879 12l-4.94 4.939 2.122 2.122L20.121 12z"
        />
        <path
          d="M6.061 19.061L13.121 12l-7.06-7.061-2.122 2.122L8.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`renders pagination with a page greater than default pages to display (5) 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: var(--button-primary-color);
  color: white;
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c1 svg {
  width: 1.2rem;
}

.c1:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c1:hover {
  opacity: 1;
}

.c4 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c4 svg {
  width: 1.2rem;
}

.c4:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c4:hover {
  opacity: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 button {
  margin-right: 0.5rem;
}

.c2 {
  --button-primary-color: #1890ff;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-label="Go to first page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.121 12l4.94-4.939-2.122-2.122L3.879 12l7.06 7.061 2.122-2.122z"
        />
        <path
          d="M17.939 4.939L10.879 12l7.06 7.061 2.122-2.122L15.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to previous page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.939 4.939L6.879 12l7.06 7.061 2.122-2.122L11.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to page 2"
      class="c1 c2"
      shape="round"
    >
      2
    </button>
    <button
      aria-label="Go to page 3"
      class="c1 c2"
      shape="round"
    >
      3
    </button>
    <button
      aria-label="Go to page 4"
      class="c1 c2"
      shape="round"
    >
      4
    </button>
    <button
      aria-label="Go to page 5"
      class="c1 c2"
      shape="round"
    >
      5
    </button>
    <button
      aria-label="Go to page 6"
      class="c4 c2"
      shape="round"
    >
      6
    </button>
    <button
      aria-label="Go to next page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.061 19.061L17.121 12l-7.06-7.061-2.122 2.122L12.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to last page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.061 4.939l-2.122 2.122L15.879 12l-4.94 4.939 2.122 2.122L20.121 12z"
        />
        <path
          d="M6.061 19.061L13.121 12l-7.06-7.061-2.122 2.122L8.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`renders pagination with a page in the middle active 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: var(--button-primary-color);
  color: white;
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c1 svg {
  width: 1.2rem;
}

.c1:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c1:hover {
  opacity: 1;
}

.c4 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c4 svg {
  width: 1.2rem;
}

.c4:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c4:hover {
  opacity: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 button {
  margin-right: 0.5rem;
}

.c2 {
  --button-primary-color: #1890ff;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-label="Go to first page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.121 12l4.94-4.939-2.122-2.122L3.879 12l7.06 7.061 2.122-2.122z"
        />
        <path
          d="M17.939 4.939L10.879 12l7.06 7.061 2.122-2.122L15.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to previous page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.939 4.939L6.879 12l7.06 7.061 2.122-2.122L11.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to page 1"
      class="c1 c2"
      shape="round"
    >
      1
    </button>
    <button
      aria-label="Go to page 2"
      class="c1 c2"
      shape="round"
    >
      2
    </button>
    <button
      aria-label="Go to page 3"
      class="c4 c2"
      shape="round"
    >
      3
    </button>
    <button
      aria-label="Go to page 4"
      class="c1 c2"
      shape="round"
    >
      4
    </button>
    <button
      aria-label="Go to page 5"
      class="c1 c2"
      shape="round"
    >
      5
    </button>
    <button
      aria-label="Go to next page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.061 19.061L17.121 12l-7.06-7.061-2.122 2.122L12.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to last page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.061 4.939l-2.122 2.122L15.879 12l-4.94 4.939 2.122 2.122L20.121 12z"
        />
        <path
          d="M6.061 19.061L13.121 12l-7.06-7.061-2.122 2.122L8.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`renders pagination with last page active 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: var(--button-primary-color);
  color: white;
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c1 svg {
  width: 1.2rem;
}

.c1:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c1:hover {
  opacity: 1;
}

.c4 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c4 svg {
  width: 1.2rem;
}

.c4:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c4:hover {
  opacity: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 button {
  margin-right: 0.5rem;
}

.c2 {
  --button-primary-color: #1890ff;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-label="Go to first page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.121 12l4.94-4.939-2.122-2.122L3.879 12l7.06 7.061 2.122-2.122z"
        />
        <path
          d="M17.939 4.939L10.879 12l7.06 7.061 2.122-2.122L15.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to previous page"
      class="c1 c2"
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.939 4.939L6.879 12l7.06 7.061 2.122-2.122L11.121 12l4.94-4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to page 1"
      class="c1 c2"
      shape="round"
    >
      1
    </button>
    <button
      aria-label="Go to page 2"
      class="c1 c2"
      shape="round"
    >
      2
    </button>
    <button
      aria-label="Go to page 3"
      class="c1 c2"
      shape="round"
    >
      3
    </button>
    <button
      aria-label="Go to page 4"
      class="c1 c2"
      shape="round"
    >
      4
    </button>
    <button
      aria-label="Go to page 5"
      class="c4 c2"
      shape="round"
    >
      5
    </button>
    <button
      aria-label="Go to next page"
      class="c1 c2"
      disabled=""
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.061 19.061L17.121 12l-7.06-7.061-2.122 2.122L12.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
    <button
      aria-label="Go to last page"
      class="c1 c2"
      disabled=""
      shape="round"
    >
      <svg
        aria-hidden="true"
        class="c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.061 4.939l-2.122 2.122L15.879 12l-4.94 4.939 2.122 2.122L20.121 12z"
        />
        <path
          d="M6.061 19.061L13.121 12l-7.06-7.061-2.122 2.122L8.879 12l-4.94 4.939z"
        />
      </svg>
    </button>
  </div>
</div>
`;
